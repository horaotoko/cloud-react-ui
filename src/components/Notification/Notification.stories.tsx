import styled from "styled-components";
import { Story, Meta } from "@storybook/react";
import { Notification, NotificationProps } from "./Notification";

export default {
  component: Notification,
  title: "components/Notification",
} as Meta;

const Wrapper = styled.div`
  > * + * {
    margin-top: 24px;
  }
`;

const Template: Story<NotificationProps> = ({ isOpen, align, ...rest }) => (
  <>
    <Wrapper>
      <Notification {...rest} isOpen={true}>
        Notification
      </Notification>
      <Notification {...rest} align="center" isOpen={true}>
        Notification align=center
      </Notification>
    </Wrapper>
  </>
);

export const Default = Template.bind({});

export const Error = Template.bind({});
Error.args = { color: "error" };

export const Warning = Template.bind({});
Warning.args = { color: "warning" };
