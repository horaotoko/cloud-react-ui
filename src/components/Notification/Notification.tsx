import { forwardRef, useEffect, useState } from "react";
import styled, { css } from "styled-components";
import { defaultProps, color } from "../../theme";

export type NotificationProps = {
  color?: "success" | "warning" | "error";
  className?: string;
  children: React.ReactNode;
  align?: "center" | "left" | "right" | "start" | "end";
  isOpen?: boolean;
  onClickClose?: () => void;
};

const white = color.white;

const StyledNotification = styled.div<
  Omit<NotificationProps, "align" | "children" | "isOpen" | "onClickClose">
>`
  ${({ color = "success", theme: { notification } }) => css`
    width: 100%;
    background-color: ${notification.backgroundColor[color]};
    opacity: 0.9;
    color: ${white};
    font-size: 13px;
  `}
`;
StyledNotification.defaultProps = defaultProps;

const CrossIcon = (): JSX.Element => (
  <svg
    width="15"
    height="15"
    viewBox="0 0 15 15"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
    aria-label="hidden"
  >
    <path
      d="M15 0.714286L14.2857 0L7.50003 6.78568L0.714355 8.70763e-06L6.93137e-05 0.714294L6.78574 7.49997L0 14.2857L0.714286 15L7.50003 8.21426L14.2858 15L15.0001 14.2857L8.21432 7.49997L15 0.714286Z"
      fill="white"
    />
  </svg>
);

const CloseButton = styled.button`
  padding: 12px;
  appearance: none;
  border: none;
  background: none;
  cursor: pointer;
`;

const Action = styled.div`
  flex: 0 0;
`;

type ContentProps = {
  children: React.ReactNode;
  align?: NotificationProps["align"];
};

const Content = styled.div<ContentProps>`
  flex: 1 1 auto;
  padding: 12px;
  text-align: ${({ align }) => align};
`;

const Inner = styled.div`
  display: flex;
  align-items: center;
`;

const Notification = forwardRef<HTMLDivElement, NotificationProps>(
  ({ children, align, isOpen = false, onClickClose, ...rest }, ref) => {
    const [isActive, setIsActive] = useState(isOpen);
    const handleCloseClick = () => {
      onClickClose && onClickClose();
      setIsActive(false);
    };

    useEffect(() => {
      setIsActive(isOpen);
    }, [isOpen]);

    if (!isActive) {
      return null;
    }

    return (
      <StyledNotification ref={ref} role="alert" {...rest}>
        <Inner>
          <Content align={align}>{children}</Content>
          <Action>
            <CloseButton
              onClick={handleCloseClick}
              aria-label="閉じる"
              type="button"
            >
              <CrossIcon />
            </CloseButton>
          </Action>
        </Inner>
      </StyledNotification>
    );
  }
);

Notification.displayName = "Notification";

export { Notification };
