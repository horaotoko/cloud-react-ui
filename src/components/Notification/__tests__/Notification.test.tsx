import { render } from "@testing-library/react";
import { Notification } from "../Notification";

describe("Notification", () => {
  it("snapshot test", () => {
    const { asFragment } = render(
      <Notification color="success">Notification</Notification>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
