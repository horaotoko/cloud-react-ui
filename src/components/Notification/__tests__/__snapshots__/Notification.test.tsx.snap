// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification should be in error colorSchema 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  background-color: #EC4949;
  opacity: 0.9;
  color: #FFFFFF;
  font-size: 13px;
}

.c4 {
  padding: 12px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  background: none;
  cursor: pointer;
}

.c3 {
  -webkit-flex: 0 0;
  -ms-flex: 0 0;
  flex: 0 0;
}

.c2 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  padding: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
    color="error"
    role="alert"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Notification
      </div>
      <div
        class="c3"
      >
        <button
          aria-label="閉じる"
          class="c4"
          type="button"
        >
          <svg
            aria-label="hidden"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 0.714286L14.2857 0L7.50003 6.78568L0.714355 8.70763e-06L6.93137e-05 0.714294L6.78574 7.49997L0 14.2857L0.714286 15L7.50003 8.21426L14.2858 15L15.0001 14.2857L8.21432 7.49997L15 0.714286Z"
              fill="white"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Notification should be in success colorSchema as default 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  background-color: #3B7DE9;
  opacity: 0.9;
  color: #FFFFFF;
  font-size: 13px;
}

.c4 {
  padding: 12px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  background: none;
  cursor: pointer;
}

.c3 {
  -webkit-flex: 0 0;
  -ms-flex: 0 0;
  flex: 0 0;
}

.c2 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  padding: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
    role="alert"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Notification
      </div>
      <div
        class="c3"
      >
        <button
          aria-label="閉じる"
          class="c4"
          type="button"
        >
          <svg
            aria-label="hidden"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 0.714286L14.2857 0L7.50003 6.78568L0.714355 8.70763e-06L6.93137e-05 0.714294L6.78574 7.49997L0 14.2857L0.714286 15L7.50003 8.21426L14.2858 15L15.0001 14.2857L8.21432 7.49997L15 0.714286Z"
              fill="white"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Notification should be in warning colorSchema 1`] = `
<DocumentFragment>
  .c0 {
  width: 100%;
  background-color: #C0AA7B;
  opacity: 0.9;
  color: #FFFFFF;
  font-size: 13px;
}

.c4 {
  padding: 12px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  background: none;
  cursor: pointer;
}

.c3 {
  -webkit-flex: 0 0;
  -ms-flex: 0 0;
  flex: 0 0;
}

.c2 {
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  padding: 12px;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

<div
    class="c0"
    color="warning"
    role="alert"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Notification
      </div>
      <div
        class="c3"
      >
        <button
          aria-label="閉じる"
          class="c4"
          type="button"
        >
          <svg
            aria-label="hidden"
            fill="none"
            height="15"
            viewBox="0 0 15 15"
            width="15"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15 0.714286L14.2857 0L7.50003 6.78568L0.714355 8.70763e-06L6.93137e-05 0.714294L6.78574 7.49997L0 14.2857L0.714286 15L7.50003 8.21426L14.2858 15L15.0001 14.2857L8.21432 7.49997L15 0.714286Z"
              fill="white"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
